lib/RequestRouter.js

```diff
@@ -0,0 +1,809 @@
+    var params = request.getParams();
+    var resource = request.getResource();
+
+    //Set the default queries in routes if not already specified in request
+    if (request.isGet()) {
+        //Set the default queries first
+        for (var key in route.query) {
+            //if getQuery is not undefined then setQuery
+            if (request.getQuery(key)) {
+                continue;
+            }
+
+            value = route.query[key];
+
+            //If there's no default values and the param has not been set then don't set it
+            if (value !== '') {
+                queries[key] = value;
+            }
+        }
+
+        //Remove queries that are not set in the route
+        for (var key in queries) {
+            if (!route.query.hasOwnProperty(key)) {
+                delete queries[key];
+            }
+        }
+    }
+
+    // Update parameters with parameters in the route url such as /test/:fooId/foo. Parameters in this case is fooId
+    for (var i = 0, len = route.keys.length; i < len; i++) {
+        var paramName = route.keys[i].name;
+        var match = this.getResourceId().match(route.regex);
+        var paramValue = match[i + 1];
+
+        if (typeof paramValue == 'undefined') {
+            paramValue = null;
+        }
+
+        params[paramName] = paramValue;
+        queries[paramName] = paramValue;
+    }
+
+    var method = this.getMethod();
+
+    if (!!~[Request.METHOD_PATCH, Request.METHOD_POST, Request.METHOD_PUT, Request.METHOD_DELETE].indexOf(method) && resource) {
+        var validate;
+
+        // Using duck-typing to see if the schema represents a collection
+        var isCollection = this.isCollection();
+
+        if (method === Request.METHOD_POST && isCollection) {
+            validate = jsen(_.extend({additionalProperties: false}, route.schema.properties.items.items));
+        } else {
+            validate = jsen(_.extend({additionalProperties: false}, route.schema));
+        }
+
+        //Remove props from resource
+        remove$Props.call(resource);
+        this.setResource(resource);
+    }
+
+    this.setParams(params);
+
+    if (request.isGet()) {
+        this.setQueries(queries);
+    }
+}
+
+// Upload validation errors
+RequestRouter.ERROR_UPLOAD_TOO_LARGE              = 'ERROR_UPLOAD_TOO_LARGE';
+RequestRouter.ERROR_UPLOAD_TOO_SMALL              = 'ERROR_UPLOAD_TOO_SMALL';
+RequestRouter.ERROR_UPLOAD_MIME_TYPE              = 'ERROR_UPLOAD_MIME_TYPE';
+RequestRouter.ERROR_UPLOAD_NOT_FOUND              = 'ERROR_UPLOAD_NOT_FOUND';
+
+var remove$Props = function(resource) {
+    if (!resource || !_.isObject(resource)) {
+        return;
+    }
+
+    _.each(_.keys(resource), function (key) {
+        if (key.indexOf('$') == 0) {
+            delete resource[key];
+        } else {
+            remove$Props(resource[key]);
+        }
+    });
+};
+
+RequestRouter.prototype.setUrl = function(url) {
+    this._request._url = url;
```

Comment:
Use a setter method, e.g. `this._request.setUrl(url)` instead of modifying the property directly. If it doesn't have this method, add it.